src/ray/common/runtime_env_manager.cc

```diff
@@ -6,15 +6,31 @@ void RuntimeEnvManager::AddUriReference(const std::string &hex_id,
                                         const rpc::RuntimeEnv &runtime_env) {
   const auto &uris = runtime_env.uris();
   for (const auto &uri : uris) {
-    if (unused_uris_.count(uri)) {
-      unused_uris_.erase(uri);
-    }
-    uri_reference_[uri]++;
-    id_to_uris_[hex_id].push_back(uri);
+    AddUriReference(hex_id, uri);
+  }
+}
+
+void RuntimeEnvManager::AddUriReference(const std::string &hex_id,
+                                        const std::string &uri) {
+  if (unused_uris_.count(uri)) {
+    unused_uris_.erase(uri);
   }
+  uri_reference_[uri]++;
+  id_to_uris_[hex_id].push_back(uri);
+}
+
+const std::vector<std::string> &RuntimeEnvManager::GetReferences(
```

Comment:
Hmm why are we returned a reference here? This seems somewhat dangerous -- should we return a shared_ptr instead?

Reply:
It's not a shared semantics. It's just GET operations. I don't think we should abuse shared pointers here. 

If the caller needs it and delegates it to other threads, the caller will just copy it and here it's not the case.